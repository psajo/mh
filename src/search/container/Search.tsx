import { Col, Row, Typography } from 'antd';
import 'antd/dist/antd.css';
import React from 'react';
import Settings from '../component/Settings';
import SearchInput from './SearchInput';

export default function Search(): JSX.Element {
  return (
    <>
      <Row justify="end" style={{ padding: 20 }}>
        <Col>
          <Settings logout={() => {}} />
        </Col>
      </Row>
      <Row justify="center" style={{ marginTop: 100 }}>
        <Col>
          <Typography.Title style={{ fontFamily: 'Caligraphy' }}>
            찾 아 야 한 다
          </Typography.Title>
        </Col>
      </Row>
      <Row justify="center" style={{ marginTop: 50 }}>
        <Col span={12}>
          <SearchInput />
        </Col>
      </Row>
    </>
  );
}
