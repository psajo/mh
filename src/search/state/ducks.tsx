import produce from 'immer';
// Actions
export const SET_KEYWORD = 'search/SET_KEYWORD';
export const SET_AUTOCOMPLETES = 'search/SET_AUTOCOMPLETES';
export const FETCH_AUTOCOMPLETES = 'search/FETCH_AUTOCOMPLETES';

export type SearchAction =
  | ReturnType<typeof setKeywordAC>
  | ReturnType<typeof setAutoCompletesAC>
  | ReturnType<typeof fetchAutoCompletesAC>;
const INITIAL_STATE: SearchState = {
  keyword: '',
  autoCompletes: []
};
export interface SearchState {
  keyword?: string;
  autoCompletes?: UserData[];
}

// Reducer - immer 사용
export default function mySearchReducer(
  state = INITIAL_STATE,
  action: SearchAction
): SearchState {
  switch (action.type) {
    case SET_KEYWORD: // 키워드 저장
      return produce(state, draft => {
        draft.keyword = action.payload.keyword;
      });
    case SET_AUTOCOMPLETES: // 자동완성 저장
      return produce(state, draft => {
        draft.autoCompletes = action.payload.autoCompletes;
      });
    default:
      return state;
  }
}

// Action Creators
export function setKeywordAC(
  keyword: string
): {
  type: typeof SET_KEYWORD;
  payload: {
    keyword: string;
  };
} {
  return { type: SET_KEYWORD, payload: { keyword } };
}

export function setAutoCompletesAC(
  autoCompletes: Array<UserData>
): {
  type: typeof SET_AUTOCOMPLETES;
  payload: {
    autoCompletes: UserData[];
  };
} {
  return { type: SET_AUTOCOMPLETES, payload: { autoCompletes } };
}

export function fetchAutoCompletesAC(
  keyword: string
): {
  type: typeof FETCH_AUTOCOMPLETES;
  payload: {
    keyword: string;
  };
} {
  return { type: FETCH_AUTOCOMPLETES, payload: { keyword } };
}
