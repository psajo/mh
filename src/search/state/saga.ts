import {
  all,
  put,
  call,
  takeEvery,
  AllEffect,
  ForkEffect
} from 'redux-saga/effects';
// import { actions, Types} from './index';
import {
  setAutoCompletesAC,
  FETCH_AUTOCOMPLETES,
  fetchAutoCompletesAC
} from './ducks';
import { callApi } from '../../common/util/api';

function* fetchAutoComplete({
  payload
}: ReturnType<typeof fetchAutoCompletesAC>) {
  const { keyword } = payload;
  const { isSuccess, data } = yield call(callApi, {
    url: '/user/search',
    params: { keyword }
  });
  if (isSuccess && data) {
    yield put(setAutoCompletesAC(data));
  }
}

export default function* searchSaga(): Generator<
  AllEffect<ForkEffect<never>>,
  void,
  unknown
> {
  yield all([takeEvery(FETCH_AUTOCOMPLETES, fetchAutoComplete)]);
}
