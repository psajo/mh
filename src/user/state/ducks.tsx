import produce from 'immer';
// Actions
export const SET_USER = 'user/SET_USER' as const;
export const FETCH_USER = 'user/FETCH_USER' as const;

export type UserActionType = typeof SET_USER | typeof FETCH_USER;
export type UserAction =
  | ReturnType<typeof setUserAC>
  | ReturnType<typeof fetchUserAC>;

export interface UserState {
  user: UserData | undefined;
}
const INITIAL_STATE = {
  user: undefined
};

// Reducer - immer 사용
export default function myUserReducer(
  state: UserState = INITIAL_STATE,
  action: UserAction
): UserState {
  switch (action.type) {
    case SET_USER:
      return produce(state, draft => {
        draft.user = action.payload.user;
      });
    default:
      return state;
  }
}

// Action Creators
export function setUserAC(
  user: UserData
): {
  type: 'user/SET_USER';
  payload: {
    user: UserData;
  };
} {
  return { type: SET_USER, payload: { user } };
}

export function fetchUserAC(
  name: string
): {
  type: 'user/FETCH_USER';
  payload: {
    name: string;
  };
} {
  return { type: FETCH_USER, payload: { name } };
}
