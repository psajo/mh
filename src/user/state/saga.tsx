import {
  all,
  AllEffect,
  call,
  ForkEffect,
  put,
  takeEvery
} from 'redux-saga/effects';
import { callApi } from '../../common/util/api';
import { fetchAutoCompletesAC } from '../../search/state/ducks';
// import { actions, Types } from '.';
import { FETCH_USER, setUserAC } from './ducks';

function* fetchUser({ payload }: ReturnType<typeof fetchAutoCompletesAC>) {
  // const name = payload.name;
  const { keyword } = payload;
  const { isSuccess, data } = yield call(callApi, {
    url: '/user/search',
    params: { keyword }
  });
  if (isSuccess && data) {
    const user = data.find((item: UserData) => item.name === keyword);
    if (user) {
      // yield put(actions.setValue('user', user));
      yield put(setUserAC(user));
    }
  }
}

export default function* userSaga(): Generator<
  AllEffect<ForkEffect<never>>,
  void,
  unknown
> {
  yield all([
    takeEvery(FETCH_USER, fetchUser)
    // takeEvery(Types.FetchUser, fetchUser)
    //   makeFetchSaga({fetchSaga: fetchUser, canCache: true}))
  ]);
}
