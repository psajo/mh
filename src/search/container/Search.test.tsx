import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import Search from './Search';
// import store from '../../common/store';
import configureStore from 'redux-mock-store';
import { UserAction } from '../../user/state/ducks';
import { SearchAction } from '../state/ducks';

jest.mock('../component/Settings', () => () => <></>);
// jest.mock('./SearchInput', () => () => <></>);

test('shoud match the snapshot when render Search Component', () => {
  //given

  //when
  const createMockStore = configureStore();
  const component = render(
    <Provider
      store={createMockStore({
        search: { keyword: '', autoCompletes: [] }
      })}
    >
      <Search />
    </Provider>
  );

  //then
  expect(component).toMatchSnapshot();
});
