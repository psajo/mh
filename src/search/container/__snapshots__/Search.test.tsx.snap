// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shoud match the snapshot when render Search Component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ant-row ant-row-end"
        style="padding: 20px;"
      >
        <div
          class="ant-col"
        />
      </div>
      <div
        class="ant-row ant-row-center"
        style="margin-top: 100px;"
      >
        <div
          class="ant-col"
        >
          <h1
            class="ant-typography"
            style="font-family: Caligraphy;"
          >
            찾 아 야 한 다
          </h1>
        </div>
      </div>
      <div
        class="ant-row ant-row-center"
        style="margin-top: 50px;"
      >
        <div
          class="ant-col ant-col-12"
        >
          <div
            class="ant-select ant-select-auto-complete ant-select-single ant-select-customize-input ant-select-show-search"
            style="width: 100%;"
          >
            <div
              class="ant-select-selector"
            >
              <span
                class="ant-select-selection-search"
              >
                <span
                  class="ant-input-group-wrapper ant-input-group-wrapper-lg ant-input-search ant-input-search-large ant-select-selection-search-input"
                >
                  <span
                    class="ant-input-wrapper ant-input-group"
                  >
                    <span
                      class="ant-input-affix-wrapper ant-input-affix-wrapper-focused ant-input-affix-wrapper-lg"
                    >
                      <span
                        class="ant-input-prefix"
                      >
                        <span
                          aria-label="search"
                          class="anticon anticon-search"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="search"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                            />
                          </svg>
                        </span>
                      </span>
                      <input
                        aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                        aria-autocomplete="list"
                        aria-controls="rc_select_TEST_OR_SSR_list"
                        aria-haspopup="listbox"
                        aria-owns="rc_select_TEST_OR_SSR_list"
                        autocomplete="off"
                        class="ant-input ant-input-lg"
                        id="rc_select_TEST_OR_SSR"
                        placeholder="검색어를 입력해주세요"
                        role="combobox"
                        type="search"
                        value=""
                      />
                    </span>
                    <span
                      class="ant-input-group-addon"
                    >
                      <button
                        class="ant-btn ant-btn-lg ant-btn-icon-only ant-input-search-button"
                        type="button"
                      >
                        <span
                          aria-label="search"
                          class="anticon anticon-search"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="search"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                            />
                          </svg>
                        </span>
                      </button>
                    </span>
                  </span>
                </span>
              </span>
              <span
                class="ant-select-selection-placeholder"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ant-row ant-row-end"
      style="padding: 20px;"
    >
      <div
        class="ant-col"
      />
    </div>
    <div
      class="ant-row ant-row-center"
      style="margin-top: 100px;"
    >
      <div
        class="ant-col"
      >
        <h1
          class="ant-typography"
          style="font-family: Caligraphy;"
        >
          찾 아 야 한 다
        </h1>
      </div>
    </div>
    <div
      class="ant-row ant-row-center"
      style="margin-top: 50px;"
    >
      <div
        class="ant-col ant-col-12"
      >
        <div
          class="ant-select ant-select-auto-complete ant-select-single ant-select-customize-input ant-select-show-search"
          style="width: 100%;"
        >
          <div
            class="ant-select-selector"
          >
            <span
              class="ant-select-selection-search"
            >
              <span
                class="ant-input-group-wrapper ant-input-group-wrapper-lg ant-input-search ant-input-search-large ant-select-selection-search-input"
              >
                <span
                  class="ant-input-wrapper ant-input-group"
                >
                  <span
                    class="ant-input-affix-wrapper ant-input-affix-wrapper-focused ant-input-affix-wrapper-lg"
                  >
                    <span
                      class="ant-input-prefix"
                    >
                      <span
                        aria-label="search"
                        class="anticon anticon-search"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="search"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                          />
                        </svg>
                      </span>
                    </span>
                    <input
                      aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                      aria-autocomplete="list"
                      aria-controls="rc_select_TEST_OR_SSR_list"
                      aria-haspopup="listbox"
                      aria-owns="rc_select_TEST_OR_SSR_list"
                      autocomplete="off"
                      class="ant-input ant-input-lg"
                      id="rc_select_TEST_OR_SSR"
                      placeholder="검색어를 입력해주세요"
                      role="combobox"
                      type="search"
                      value=""
                    />
                  </span>
                  <span
                    class="ant-input-group-addon"
                  >
                    <button
                      class="ant-btn ant-btn-lg ant-btn-icon-only ant-input-search-button"
                      type="button"
                    >
                      <span
                        aria-label="search"
                        class="anticon anticon-search"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="search"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                          />
                        </svg>
                      </span>
                    </button>
                  </span>
                </span>
              </span>
            </span>
            <span
              class="ant-select-selection-placeholder"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
